TechnicalDocuments/Proposals/StyleSheets.md

```diff
@@ -0,0 +1,119 @@
+
+    On the other hand, when one defines a generic `ExtensibleViewStyleSheet<T: ExtensibleView>` stylesheet, it does address the said issue by satisfying `StyleSheetProtocol.Element` with a generic parameter. However, it leads to awkward and verbose spellings at use sites in general, e.g. `ExtensibleViewStyleSheet<ExtensibleView>`.
+
+3. Correctness: Unintended Sharing
+
+    Stylesheets are currently classes so as to be able to take advantage of inheritance, and this makes them prone to unintended sharing. With Swift offering first-class value type, this might amplify its possibility due to common perception of configurations being modelled as value types.
+
+## Identified Requirements
+
+1. The approach must be applicable to reusable views.
+
+    When multiple style sheets with symmetrical differences are applied in a row in any order, the reused view would not end up in an indeterministic state.
+
+1. The approach must be compatible with view subclassing.
+
+1. The approach should address unintended sharing if possible.
+
+## Proposed Solution
+
+A new value type `StyleSheet<View>` shall be introduced, with the following features:
+
+1. Copy on write & value semantics: Backed by CoW `Dictionary`. Support `Equatable` out of the box.
+
+2. Built around Key Paths introduced since Swift 4, and has proper support of entries setting key paths that are partially overlapping with each other.
+
+3. Support producing **inverses** for deterministic view reuse.
+
+    Before applying a stylesheet, a snapshot of all affected key paths is taken, and is returned to the callee after the application. It is up to the callee whether to and when to apply the snapshot for change reversal.
+
+    In the case of a reusable view/cell, the expectation is that the view would store the inverse/snapshot & apply it as part of the cleanup e.g. in `prepareForReuse()`.
+    
+    This eliminates the state indeterminism of closure-based styling approaches, while not requiring any default value to be declared upfront like how our `StyleSheet` framework currently does. 
+
+4. Can be used for any view and any arbitrary subclass without type declaration.
+
+### Implementation Plan
+
+1. As soon as the new `StyleSheet` type is available in Bento, new components shall start using it.
+
+2. Extensions to either `UIView` or `Base*View` shall be made to provide a convenience which stores & applies style sheets on one's behalf. For example:
+
+   ```swift
+   view.backgroundColor = .green
+   view.clipsToBound = false
+
+    // NOTE: `apply` automatically keeps the inverse.
+   view.apply(
+       StyleSheet()
+            .setting(\.backgroundColor, .red)
+            .setting(\.clipsToBound, true)
+    )
+   expect(view.backgroundColor) == .red
+   expect(view.clipsToBound) == true
+
+    // NOTE: The stored inverse is applied before the new style sheet is applied.
+    //       So `clipsToBound` would be reverted to the original value.
```

Comment:
Can you update the code above to _manually_ apply inverted style that is described in https://github.com/Babylonpartners/Bento/pull/88 ?

Reply:
Sure. But in any case I needed to rewrite part of the proposal. Just didn't have time to do it...